### PR TITLE
Support for Http Proxy Headers

### PR DESCRIPTION
**Pull request by KarstenSchnitter**
This PR addresses several http headers, that may be generated by reverse proxies for incoming http requests. They are added to the request logs generated by the library, provided that certain environment variables are set. See the commit messages for details. The documentation will be changed accordingly.

### CODE REVIEW

File: cf-java-logging-support-core/beats/request-metrics/docs/fields.asciidoc [hunk 1]
```diff
@@ -695,3 +695,132 @@ client, followed by proxy server addresses that forwarded the client
...
+
+A header set by HA-Proxy to provide the end date of the client
+certificate as a formatted string YYMMDDhhmmss.
+
+
+==== x_ssl_client_session_id
+
+type: string
+
+example: session-id
```
Comment (juergen-walter) [inline cf-java-logging-support-core/beats/request-metrics/docs/fields.asciidoc:820]: Isn't the session id usually a hex value with predefined length?

File: cf-java-logging-support-core/beats/request-metrics/docs/fields.asciidoc [hunk 2]
```diff
@@ -695,3 +695,132 @@ client, followed by proxy server addresses that forwarded the client
 request.
 
 
+==== x_forwarded_host
+
+type: string
+
+example: requested-host.example.com
+
+required: False
```
Comment (juergen-walter) [inline cf-java-logging-support-core/beats/request-metrics/docs/fields.asciidoc:704]: What is the reason to have "False" starting with capital letter in this document?

File: cf-java-logging-support-core/src/main/java/com/sap/hcp/cf/logging/common/request/RequestRecordBuilder.java [hunk 3]
```diff
@@ -39,7 +39,7 @@ public RequestRecordBuilder addOptionalTag(boolean optionalFieldCanBeLogged, Str
 			requestRecord.addTag(fieldKey, Defaults.REDACTED);
 		}
 
-		if (!optionalFieldCanBeLogged && tag.equals(Defaults.UNKNOWN)) {
+        if (!optionalFieldCanBeLogged && Defaults.UNKNOWN.equals(tag)) {
```
Comment (juergen-walter) [inline cf-java-logging-support-core/src/main/java/com/sap/hcp/cf/logging/common/request/RequestRecordBuilder.java:42]:
```suggestion
		if (!optionalFieldCanBeLogged && Defaults.UNKNOWN.equals(tag)) {
```

Comment (juergen-walter) [reply cf-java-logging-support-core/src/main/java/com/sap/hcp/cf/logging/common/request/RequestRecordBuilder.java:42]: fix indentation

File: cf-java-logging-support-core/beats/request-metrics/docs/fields.asciidoc (hunk 2 above)
Comment (KarstenSchnitter) [reply cf-java-logging-support-core/beats/request-metrics/docs/fields.asciidoc:704]: It is auto-generated. Always has been like that.

**Comment by juergen-walter** [comment]
Please align indentation method in your PR: 

RequestLoggingFilterTest.java uses spaces for indentation

RequestRecordBuilder.java uses tabs

File: cf-java-logging-support-core/beats/request-metrics/docs/fields.asciidoc (hunk 1 above)
Comment (KarstenSchnitter) [reply cf-java-logging-support-core/beats/request-metrics/docs/fields.asciidoc:820]: This is auto-generated from the yaml. Anyways, there is no guessing, what kind of ids users of the library may use. Could also be uuids or usernames including timestamps. Not that I would advise doing something like that.

File: cf-java-logging-support-core/src/main/java/com/sap/hcp/cf/logging/common/request/RequestRecordBuilder.java (hunk 3 above)
Comment (KarstenSchnitter) [reply cf-java-logging-support-core/src/main/java/com/sap/hcp/cf/logging/common/request/RequestRecordBuilder.java:42]: I reformatted the whole class with the correct formatting rules. It was once accidentally formatted wrongly by me.

**Comment by KarstenSchnitter** [comment]
The usual indentation should be using spaces as defined in https://github.com/SAP/cf-java-logging-support/blob/master/eclipse-formatter-settings.xml. Unfortunately, I once formatted several files using tabs. I reformatted the RequestRecordBuilder as a whole. I do not want to run the formatter over the whole repository, since this breaks file history too much for me.

**Comment by juergen-walter** [comment]
> The usual indentation should be using spaces as defined in https://github.com/SAP/cf-java-logging-support/blob/master/eclipse-formatter-settings.xml. Unfortunately, I once formatted several files using tabs. I reformatted the RequestRecordBuilder as a whole. I do not want to run the formatter over the whole repository, since this breaks file history too much for me.

fine for me., thanks for the pointer. Just got confused by different indentation methods.

File: cf-java-logging-support-core/beats/request-metrics/etc/fields.yml [hunk 4]
```diff
@@ -501,3 +501,88 @@ request-metrics:
         Comma-separated list of IP addresses, the left-most being the original
         client, followed by proxy server addresses that forwarded the client
         request.
+
+    - name: "x_forwarded_host"
+      type: string
+      required: false
+      example: "requested-host.example.com"
+      description: |
+        The originally reqeusted host by the client in the host HTTP request header.
```
Comment (christiand93) [inline cf-java-logging-support-core/beats/request-metrics/etc/fields.yml:510]:
```suggestion
        The originally requested host by the client in the host HTTP request header.
```

File: cf-java-logging-support-core/beats/request-metrics/etc/fields.yml [hunk 5]
```diff
@@ -501,3 +501,88 @@ request-metrics:
...
+      required: false
+      example: "/C=FR/ST=Ile de France/L=Jouy en Josas/O=haproxy.com/CN=haproxy.com/emailAddress=ba@haproxy.com"
+      description: |
+        A header set by HA-Proxy to provide the full distinguished name of the
+        issuer of the client certificate.
+
+    - name: "x_ssl_client_notbefore"
+      type: string
+      required: false
+      example: "/C=FR/ST=Ile de France/L=Jouy en Josas/O=haproxy.com/CN=haproxy.com/emailAddress=ba@haproxy.com"
```
Comment (christiand93) [inline cf-java-logging-support-core/beats/request-metrics/etc/fields.yml:570]: The example does not show a date string.

File: cf-java-logging-support-core/beats/request-metrics/etc/fields.yml [hunk 6]
```diff
@@ -501,3 +501,88 @@ request-metrics:
...
+      description: |
+        A header set by HA-Proxy to provide the start date of the client
+        certificate as a formatted string YYMMDDhhmmss.
+
+    - name: "x_ssl_client_notafter"
+      type: string
+      required: false
+      example: "/C=FR/ST=Ile de France/L=Jouy en Josas/O=haproxy.com/CN=haproxy.com/emailAddress=ba@haproxy.com"
```
Comment (christiand93) [inline cf-java-logging-support-core/beats/request-metrics/etc/fields.yml:578]: same as above.

**Review by christiand93 — APPROVED** [review]
LGTM 👍

**Pull request merged**